libxrdp/xrdp_sec.c

```diff
@@ -2302,159 +2302,235 @@ xrdp_sec_process_mcs_data_channels(struct xrdp_sec *self, struct stream *s)
-            y1 = MIN(y1, client_info->minfo[index].top);
-            x2 = MAX(x2, client_info->minfo[index].right);
-            y2 = MAX(y2, client_info->minfo[index].bottom);
+            /*
+             * 2.2.1.3.6.1 Monitor Definition (TS_MONITOR_DEF)
+             */
+            LOG(LOG_LEVEL_TRACE, "\tprocess_monitor_stream: "
+                "Received [MS-RDPBCGR] 2.2.1.3.6.1 TS_UD_CS_MONITOR.TS_MONITOR_DEF "
+                "Index: %d, Left %d, Top %d, Right %d, Bottom %d, Flags 0x%8.8x",
+                monitor_index,
+                monitor_layout->left,
+                monitor_layout->top,
+                monitor_layout->right,
+                monitor_layout->bottom,
+                monitor_layout->is_primary);
         }
 
-        if (client_info->minfo[index].is_primary)
+        if (monitor_index == 0)
         {
-            got_primary = 1;
+            rect.left = monitor_layout->left;
+            rect.top = monitor_layout->top;
+            rect.right = monitor_layout->right;
+            rect.bottom = monitor_layout->bottom;
+        }
+        else
+        {
+            rect.left = MIN(monitor_layout->left, rect.left);
+            rect.top = MIN(monitor_layout->top, rect.top);
+            rect.right = MAX(rect.right, monitor_layout->right);
+            rect.bottom = MAX(rect.bottom, monitor_layout->bottom);
         }
 
-        LOG(LOG_LEVEL_DEBUG,
-            "Client monitor [%d]: left= %d, top= %d, right= %d, bottom= %d, "
-            "is_primary?= %d",
-            index,
-            client_info->minfo[index].left,
-            client_info->minfo[index].top,
-            client_info->minfo[index].right,
-            client_info->minfo[index].bottom,
-            client_info->minfo[index].is_primary);
+        if (monitor_layout->is_primary)
+        {
+            got_primary = 1;
+        }
     }
 
     if (!got_primary)
     {
         /* no primary monitor was set, choose the leftmost monitor as primary */
-        for (index = 0; index < monitorCount; index++)
+        for (int monitor_index = 0; monitor_index < num_monitor; ++monitor_index)
         {
-            if (client_info->minfo[index].left == x1 &&
-                    client_info->minfo[index].top == y1)
+            monitor_layout = description->minfo + monitor_index;
+            if (monitor_layout->left != rect.left || monitor_layout->top != rect.top)
             {
-                client_info->minfo[index].is_primary = 1;
-                break;
+                continue;
             }
+            monitor_layout->is_primary = 1;
         }
     }
 
     /* set wm geometry */
-    if ((x2 > x1) && (y2 > y1))
+    if ((rect.right > rect.left) && (rect.bottom > rect.top))
     {
-        client_info->width = (x2 - x1) + 1;
-        client_info->height = (y2 - y1) + 1;
+        description->session_width = rect.right - rect.left;
+        description->session_height = rect.bottom - rect.top;
     }
```

Comment:
Good point -- I added the error.

The +1 was removed because it wasn't present in the similar code that came from the resizing code, but it probably should be there as it's actually more correct.